src/script/interpreter.h

```diff
@@ -193,6 +193,29 @@ enum class SigVersion
     TAPSCRIPT = 3,   //!< Witness v1 with 32-byte program, not BIP16 P2SH-wrapped, script path spending, leaf version 0xc0; see BIP 342
 };
 
+
+//! Data that is accumulated during the script verification of a single input and then
+//! used to perform aggregate checks after all inputs have been run through
+//! `VerifyScript()`.
+//!
+//! TODO I realize this is kind of a hacky polymorphic-union thing; if there's a nicer
+//! way to do this, please someone let me know.
+struct DeferredCheck
+{
+    //! Set when script execution happens asynchronously so that we can associate
+    //! deferred checks with their related transaction when a block's worth of
+    //! script executions are performed in batch.
+    const CTransaction* m_tx_to;
```

Comment:
I think the way you're setting this up is that you give `CCheckQueue` a bunch of work, and it then returns a batch of deferred work back to you, that you then finish off. But if you're doing a whole block at once, maybe that's bad, because you're doing all the deferred checks for all txns in a single thread, rather than in parallel.

Maybe it would be better to expand `CCheckQueue` into being more of a map/reduce model -- you map the scripts to many processes getting deferred checks in response, then you sort the deferred checks so related ones (from the same tx) all go to the same thread, then reduce/accumulate the results from them so that you've got your conclusion.

Could be really clever and do the reduce work in parallel with mapping, I think; though you'd need some way of maintaining the accumulator state.

Reply:
This is good feedback, and in the general case you're right. In the specific case of the deferred checks being introduced here, they don't include any of the computationally expensive stuff that motivates parallelizing checks in the case of CScriptCheck - i.e. we're not doing any cryptographic ops or signature checking here; OP_VAULT deferred checks consist of summing and comparing integers.

So I'm not necessarily averse to doing the deferred-check parallelization in this changeset, but (i) I don't think it's strictly necessary from a resource standpoint and (ii) I can't think of a straightforward way to do it without introducing another thread pool devoted to the deferred checks, which is a decent amount of work.